tools/gfx/cuda/render-cuda.cpp

```diff
@@ -0,0 +1,1506 @@
+    }
+
+    virtual void unmap(BufferResource* buffer) override
+    {
+        SLANG_UNUSED(buffer);
+    }
+
+    virtual void setPipelineState(PipelineType pipelineType, PipelineState* state) override
+    {
+        SLANG_ASSERT(pipelineType == PipelineType::Compute);
+        currentPipeline = dynamic_cast<CUDAPipelineState*>(state);
+    }
+
+    virtual void dispatchCompute(int x, int y, int z) override
+    {
+        // Find out thread group size from program reflection.
+        auto& kernelName = currentPipeline->shaderProgram->kernelName;
+        auto programLayout = dynamic_cast<CUDAProgramLayout*>(currentRootObject->layout.Ptr());
+        int kernelId = programLayout->getKernelIndex(kernelName.getUnownedSlice());
+        SLANG_ASSERT(kernelId != -1);
+        UInt threadGroupSize[3];
+        programLayout->getKernelThreadGroupSize(kernelId, threadGroupSize);
+
+        int sharedSizeInBytes;
+        cuFuncGetAttribute(
+            &sharedSizeInBytes,
+            CU_FUNC_ATTRIBUTE_SHARED_SIZE_BYTES,
+            currentPipeline->shaderProgram->cudaKernel);
+
+        // Copy global parameter data to the `SLANG_globalParams` symbol.
+        {
+            CUdeviceptr globalParamsSymbol = 0;
+            size_t globalParamsSymbolSize = 0;
+            cuModuleGetGlobal(
+                &globalParamsSymbol,
+                &globalParamsSymbolSize,
+                currentPipeline->shaderProgram->cudaModule,
+                "SLANG_globalParams");
+
+            CUdeviceptr globalParamsCUDAData =
+                currentRootObject->bufferResource
+                    ? currentRootObject->bufferResource->getBindlessHandle()
+                    : 0;
+            cudaMemcpyAsync(
+                (void*)globalParamsSymbol,
+                (void*)globalParamsCUDAData,
+                globalParamsSymbolSize,
+                cudaMemcpyDeviceToDevice,
+                0);
+        }
+        //
+        // In the case of the entry-point parameters, we have to deal with
+        // two different wrinkles.
+        //
+        // First, the `bindRoot` will have the entry-point argument data
+        // stored in a GPU-memory buffer, but we actually need it to be
+        // in host CPU memory. We handle that for now by allocating a
+        // temporary host memory buffer (if needed) and copying the data
+        // from device to host.
+        //
+        auto entryPointBuffer = currentRootObject->entryPointObjects[kernelId]->bufferResource.Ptr();
+        size_t entryPointDataSize = entryPointBuffer ? entryPointBuffer->getDesc().sizeInBytes : 0;
+        void* entryPointHostData = nullptr;
+        if (entryPointDataSize)
+        {
+            entryPointHostData = alloca(entryPointDataSize);
+            cudaMemcpy(
+                entryPointHostData,
+                (void*)entryPointBuffer->getBindlessHandle(),
+                entryPointDataSize,
+                cudaMemcpyDeviceToHost);
+        }
```

Comment:
Note that one option we could consider now that `ShaderObject` is under the API boundary is to have two cases of CUDA `ShaderObject`s: an ordinary `CUDAShaderObject` for the 99% case and a `CUDAEntryPointShaderObject` to handle the entry-point case. The latter would differ only in that it stores its data in host memory instead of device memory.

By having a distinct type for CUDA entry point shader objects, we could avoid the allocation and device->host copy here.